formats/json/commonMain/src/kotlinx/serialization/json/internal/StreamingJsonDecoder.kt

```diff
@@ -101,6 +101,19 @@ internal open class StreamingJsonDecoder(
         }
     }
 
+    private fun SerialDescriptor.getJsonElementIndex(key: String): Int {
+        val index = this.getElementIndex(key)
+        if (!json.configuration.useAlternativeNames) return index
+        // Fast path, do not go through ConcurrentHashMap.get
+        // Note, it blocks ability to detect collisions between the primary name and alternate,
+        // but it eliminates a significant performance penalty (about -15% without this optimization)
+        if (index != UNKNOWN_NAME) return index
+        // slow path
+        val alternativeNamesMap =
```

Comment:
We expect this map to be pretty small (1-4 entries).

So I see two more potential ways to do that lookup:
1) Do not cache anything, lookup annotation and then do a linear scan in the array.
2) Cache array of names from annotation and do a linear lookup

For our constraints one of these approaches might be slightly faster, could you please try it?